### PR TITLE
Add standalone parallel CNN functions for MNIST

### PR DESCRIPTION
**Pull request by shuds13**
For comparison with libE.

Works in parallel on Perlmutter and uses GPUs on one node, but does not work across nodes.

Main loop

```python

    for epoch in range(epochs):
        for batch_idx, (data, target) in enumerate(train_loader):
            data, target = data.to(device), target.to(device)

            optimizer.zero_grad()
            output = model(data)
            loss = nn.CrossEntropyLoss()(output, target)
            loss.backward()

            # Synchronize gradients across processes (libE would do this)
            for param in model.parameters():
                dist.all_reduce(param.grad, op=dist.ReduceOp.SUM)
                param.grad /= world_size

            optimizer.step()

```

### CODE REVIEW

File: cnn_comparator/simf_cnn_NCCL_parallel.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                    grad_shape = param.grad.shape
+                    grad_size = param.grad.numel()
+
+                    # SH - speculatively called "grads_in"
+                    # Extract the corresponding section and reshape
+                    param.grad.data.copy_(
+                        torch.tensor(calc_in["grads_in"][offset:offset + grad_size].reshape(grad_shape),
+                                    device=param.grad.device)
+                    )
+                    offset += grad_size
```
Comment (shuds13) [inline cnn_comparator/simf_cnn_NCCL_parallel.py:89]: If need to average above, then need to do it here also.

File: cnn_comparator/simf_cnn_NCCL_parallel.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            loss = nn.CrossEntropyLoss()(output, target)
+            loss.backward()
+
+            # Synchronize gradients across processes
+            for param in model.parameters():
+                dist.all_reduce(param.grad, op=dist.ReduceOp.SUM)
+                param.grad /= group_size  # SH averaging?
+
+            # SH TODO - consider inheriting dist (torch.distributed)
+            # and overriding dist.all_reduce() to do libE stuff
```
Comment (shuds13) [inline cnn_comparator/simf_cnn_NCCL_parallel.py:67]: torch.distributed is a module not a class, but could extract these lines to some provided function.

File: cnn_comparator/pytorch_cnn_NCCL_parallel.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    for epoch in range(epochs):
+        for batch_idx, (data, target) in enumerate(train_loader):
+            data, target = data.to(device), target.to(device)
+
+            optimizer.zero_grad()
+            output = model(data)
+            loss = nn.CrossEntropyLoss()(output, target)
+            loss.backward()
+
+            # Synchronize gradients across processes
```
Comment (shuds13) [inline cnn_comparator/pytorch_cnn_NCCL_parallel.py:59]: #if do optimizer step on gen gen does

optimizer.zero_grad()
gradients.sum()
optimizer.step()
